README.md
# Spotify-Control

This is a simple Perl interface for controlling a local instance of Spotify. I've only tested it on OS X but it doesn't use anything too specialised so should work fine on Linux and BSD. Windows is untested.

## Installation

Installation is straight forward using cpanm

```bash
cpanm -i git@github.com:ccakes/p5-spotify-control.git
```

## Usage

```perl
use Spotify::Control;

my $spot = Spotify::Control->new;
say $spot->status->{playing} ? "Currently playing" : "Currently paused";
$spot->play(uri => 'spotify:track:5bj4hb0QYTs44PDiwbI5CS');
```

## Contributing

The main items on the list is an Applescript and D-BUS interface. Following that, a standard format for returning the status across all the implementations, possibly as an object.

- [ ] Discover client HTTP port
- [ ] Spotify::Control::Applescript
- [ ] Spotify::Control::DBUS
- [ ] Standardise ->status response


CHANGELOG
v0.10 - Initial release
 - Basic HTTP interaction with client


lib/Spotify/Control.pm
package Spotify::Control;

use 5.006;
use strict;
use warnings FATAL => 'all';

use Carp;

use Spotify::Control::HTTP;

=head1 NAME

Spotify::Control - The great new Spotify::Control!

=head1 VERSION

Version 0.10

=cut

our $VERSION = '0.10';


=head1 SYNOPSIS

Spotify::Control::HTTP - Control a local instance of the Spotify app

Currently this only support controlling using the built-in HTTP interface however Applescript and D-BUS support is planned as well.

    use Spotify::Control;

    my $spotify = Spotify::Control->new;
    $spotify->pause; # pause current track
    $spotify->unpause; # resume current track

=head1 SUBROUTINES/METHODS

=head2 new

=cut

sub new {
    my $self = bless {}, shift;
    return unless @_ % 2 == 0;

    my %args = @_;

    my %defaults = (
        _s => undef,

        method => 'http'
    );

    foreach (keys %defaults) {
        $self->{$_} = exists $args{$_} ? $args{$_} : $defaults{$_};
    }

    if ($self->{method} eq 'http') {
        $self->{_s} = Spotify::Control::HTTP->new;
        $self->{_s}->initialise;
    }
    elsif ($self->{method} eq 'applescript') {
        croak __PACKAGE__ . '->new: Applescript not yet supported';
    }
    elsif ($self->{method} eq 'dbus') {
        croak __PACKAGE__ . '->new: D-BUS not yet supported';
    }

    return $self;
}

=head2 ua

Get or set UserAgent object

    say ref($spotify->ua);
    my $ua = my $lwp = LWP::UserAgent->new( ssl_opts => { verify_hostname => 0 } );
    $ua->proxy('https', 'http://127.0.0.1:8080');
    $spotify->ua($ua);

=cut

sub ua {
    ( ref $_[1] ) ? shift->{_ua} = $_[1] : shift->{_ua};
}

=head2 spot

Get or set an instance of the child object used to interact with the player. Any object could be used here provided it conforms to the same API

=cut

sub spot {
    ( ref $_[1] ) ? shift->{_s} = $_[1] : shift->{_s};
}

=head2 play

Play a Spotify URI. This can be a track, album or artist

    $spotify->play(uri => 'spotify:track:39XkO83cD29UMRwiAUyT1v');

=cut

sub play {
    my $self = shift;
    croak __PACKAGE__ . '->play: Invalid arguments' unless @_ % 2 == 0;

    return $self->spot->play(@_);
}

=head2 pause

Pause the player

=cut

sub pause {
    return shift->spot->pause;
}

=head2 unpause

Resume the player

=cut

sub unpause {
    return shift->spot->unpause;
}

=head2 status

Fetch the status of the player, return a hashref containing track info and activity state.

=cut

sub status {
    return shift->spot->status;
}

=head1 AUTHOR

Cameron Daniel, C<< <cdaniel at nurve.com.au> >>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Spotify::Control

=cut

1;


lib/Spotify/Control/HTTP.pm
package Spotify::Control::HTTP;

use 5.006;
use strict;
use warnings FATAL => 'all';

use Carp;
use URI;
use JSON::XS;
use LWP::UserAgent;

use Data::Dumper::Simple;

=head1 NAME

Spotify::Control::HTTP - Control your Spotify app using it's built-in HTTP interface

=head1 VERSION

Version 0.10

=cut

our $VERSION = '0.10';


=head1 SYNOPSIS

This module provides simple control of a local instance of the Spotify app using it's built-in HTTP interface

=head1 SUBROUTINES/METHODS

=head2 new

=cut

sub new {
    my $self = bless {}, shift;
    return unless @_ % 2 == 0;

    my %args = @_;

    my %defaults = (
        _ua => LWP::UserAgent->new(agent => __PACKAGE__ . '/' . $VERSION),
        _oauth => undef,
        _csrf => undef,

        hostname => join('', map(sprintf("%x", rand 16), 1..8)) . '.spotilocal.com',
        port => 4370,

        headers => ['Origin', 'https://open.spotify.com'],
    );

    foreach (keys %defaults) {
        $self->{$_} = exists $args{$_} ? $args{$_} : $defaults{$_};
    }

    return $self;
}

=head2 initialise

Grab OAuth and CSRF tokens to use in subsequent requests

    $spotify->initialise

=cut

sub initialise {
    my $self = shift;

    my $response;

    # Grab token from Spotify
    $response = $self->ua->get('http://open.spotify.com/token');
    croak __PACKAGE__ . '->initialise: ' . $response->decoded_content unless $response->is_success;

    eval { $self->{_oauth} = decode_json($self->ua->get('http://open.spotify.com/token')->content)->{t} };
    eval { $self->{_csrf} = decode_json($self->ua->get(sprintf('https://%s:%d', $self->{hostname}, $self->{port}) . '/simplecsrf/token.json', @{$self->{headers}})->content)->{token} };

    croak __PACKAGE__ . '->initialise: ' . $@ if $@;

    return 1 unless ($self->{_oauth} && $self->{_csrf});
}

=head2 ua

Get or set UserAgent object

    say ref($spotify->ua);
    my $ua = my $lwp = LWP::UserAgent->new( ssl_opts => { verify_hostname => 0 } );
    $ua->proxy('https', 'http://127.0.0.1:8080');
    $spotify->ua($ua);

=cut

sub ua {
    ( ref $_[1] ) ? shift->{_ua} = $_[1] : shift->{_ua};
}

=head2 play

Play a Spotify URI. This can be a track, album or artist

    $spotify->play(uri => 'spotify:track:39XkO83cD29UMRwiAUyT1v');

=cut

sub play {
    my $self = shift;
    return unless @_ % 2 == 0;
    my %args = @_;

    croak __PACKAGE__ . '->play: Missing URI' unless exists $args{uri};
    $args{context} = $args{uri} unless exists $args{context}; # add context if not set

    return $self->_request('/remote/play.json', %args);
}

=head2 pause

Pause the player

=cut

sub pause {
    return decode_json(shift->_request('/remote/pause.json', pause => 'true')->content);
}

=head2 unpause

Resume the player

=cut

sub unpause {
    return decode_json(shift->_request('/remote/pause.json', pause => 'false')->content);
}

=head2 status

Fetch the status of the player, return a hashref containing track info and activity state.

=cut

sub status {
    return decode_json(shift->_request('/remote/status.json', returnafter => 1)->content);
}

=head2 _request

Build HTTP request

=cut

sub _request {
    my $self = shift;
    my $uri = shift;
    return unless @_ % 2 == 0;

    my %args = @_;

    my $params = {
        oauth => $self->{_oauth},
        csrf => $self->{_csrf}
    };
    map { $params->{$_} = $args{$_} } keys %args;

    my $url = URI->new(sprintf('https://%s:%d%s', $self->{hostname}, $self->{port}, $uri));
    $url->query_form($params);

    my $response = $self->ua->get($url->as_string, @{$self->{headers}});

    return $response;
}

=head1 AUTHOR

Cameron Daniel, C<< <cdaniel at nurve.com.au> >>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

perldoc Spotify::Control::HTTP

=cut

1;
